ql/src/java/org/apache/hadoop/hive/ql/parse/MergeSemanticAnalyzer.java

```diff
@@ -350,26 +352,51 @@ private MergeStatement.InsertClause handleInsert(ASTNode whenNotMatchedClause, S
       throw new SemanticException(String.format("Column schema must have the same length as values (%d vs %d)",
           columnListNode.getChildCount(), valuesNode.getChildCount() - 1));
     }
-    UnparseTranslator defaultValuesTranslator = new UnparseTranslator(conf);
-    defaultValuesTranslator.enable();
-    List<String> targetSchema = processTableColumnNames(columnListNode, targetTable.getFullyQualifiedName());
-    collectDefaultValues(valuesNode, targetTable, targetSchema, defaultValuesTranslator);
-    defaultValuesTranslator.applyTranslations(ctx.getTokenRewriteStream());
-    String valuesClause = getMatchedText(valuesNode);
-    valuesClause = valuesClause.substring(1, valuesClause.length() - 1); //strip '(' and ')'
 
-    String extraPredicate = getWhenClausePredicate(whenNotMatchedClause);
-    return new MergeStatement.InsertClause(
-        getMatchedText(columnListNode), valuesClause, onClausePredicate, extraPredicate);
-  }
+    List<String> columnNames;
+    if (columnListNode != null) {
+      columnNames = new ArrayList<>(valuesNode.getChildCount());
+      for (int i = 0; i < columnListNode.getChildCount(); ++i) {
+        ASTNode columnNameNode = (ASTNode) columnListNode.getChild(i);
+        String columnName = ctx.getTokenRewriteStream().toString(columnNameNode.getTokenStartIndex(),
+            columnNameNode.getTokenStopIndex()).trim();
+        columnNames.add(columnName);
+      }
+    } else {
+      columnNames = null;
```

Comment:
should it be an empty list? do we handle nulls later on

Reply:
When column names are specified additional steps must be done in both Rewriter implementations so the null check is required. If the lack of column names is represented by an empty list `List.isEmpty()` would be used.

Reply:
👍